src/plans/hessian_plan.jl

```diff
@@ -405,60 +456,343 @@ function ApproxHessianFiniteDifference(
-    (norm_X ≈ zero(norm_X)) && return zero_vector!(M, X, x)
+
+function (f::ApproxHessianFiniteDifference{MutatingEvaluation})(M, Y, p, X)
+    norm_X = norm(M, p, X)
+    (norm_X ≈ zero(norm_X)) && return zero_vector!(M, X, p)
     c = f.steplength / norm_X
-    f.gradient!!(M, f.grad_tmp, x)
-    retract!(M, f.x_dir, x, c * X, f.retraction_method)
-    f.gradient!!(M, f.grad_tmp_dir, f.x_dir)
+    f.gradient!!(M, f.grad_tmp, p)
+    retract!(M, f.p_dir, p, c * X, f.retraction_method)
+    f.gradient!!(M, f.grad_tmp_dir, f.p_dir)
     vector_transport_to!(
-        M, f.grad_tmp_dir, f.x_dir, f.grad_tmp_dir, x, f.vector_transport_method
+        M, f.grad_tmp_dir, f.p_dir, f.grad_tmp_dir, p, f.vector_transport_method
     )
     Y .= (1 / c) .* (f.grad_tmp_dir .- f.grad_tmp)
     return Y
 end
 
+@doc raw"""
+    ApproxHessianSymmetricRankOne{E, P, G, T, B<:AbstractBasis{ℝ}, VTR, R<:Real}
+
+A functor to approximate the Hessian by the symmetric rank one update.
+
+# Fields
+
+* `gradient!!` the gradient function (either allocating or mutating, see `evaluation` parameter).
+* `ν` a small real number to ensure that the denominator in the update does not become too small and thus the method does not break down.
+* `vector_transport_method` a vector transport to use.
+
+## Internal temporary fields
+
+* `p_tmp` a temporary storage the current point `p`.
+* `grad_tmp` a temporary storage for the gradient at the current `p`.
+* `matrix` a temporary storage for the matrix representation of the approximating operator.
+* `basis` a temporary storage for an orthonormal basis at the current `p`.
+
+# Constructor
+
+    ApproxHessianSymmetricRankOne(M, p, gradF; kwargs...)
+
+## Keyword arguments
+
+* `initial_operator` (`Matrix{Float64}(I, manifold_dimension(M), manifold_dimension(M))`) the matrix representation of the initial approximating operator.
+* `basis` (`DefaultOrthonormalBasis()`) an orthonormal basis in the tangent space of the initial iterate p. 
+* `nu` (`-1`) 
+* `evaluation` (`AllocatingEvaluation()`) whether the gradient is given as an allocation function or an in-place (`MutatingEvaluation()`).
+* `vector_transport_method` (`ParallelTransport()`) vector transport ``\mathcal T_{\cdot\gets\cdot}`` to use.
+"""
+mutable struct ApproxHessianSymmetricRankOne{E,P,G,T,B<:AbstractBasis{ℝ},VTR,R<:Real}
+    p_tmp::P
+    gradient!!::G
+    grad_tmp::T
+    matrix::Matrix
+    basis::B
+    vector_transport_method::VTR
+    ν::R
+end
+function ApproxHessianSymmetricRankOne(
+    M::mT,
+    p::P,
+    gradient::G;
+    initial_operator::AbstractMatrix=Matrix{Float64}(
+        I, manifold_dimension(M), manifold_dimension(M)
+    ),
+    basis::B=DefaultOrthonormalBasis(),
+    nu::R=-1.0,
+    evaluation=AllocatingEvaluation(),
+    vector_transport_method::VTR=ParallelTransport(),
+) where {
+    mT<:AbstractManifold,P,G,B<:AbstractBasis{ℝ},R<:Real,VTR<:AbstractVectorTransportMethod
+}
+    if evaluation == AllocatingEvaluation()
```

Comment:
What is the function`gradient` here? Also please do not do this with an `if` but with multiple dispatch i.e. by implementing two functions. We could do a function `get_gradient` (we have that for problems and options already but could extend that to a `(M,grad,eval,p)` signature as well.